README.rst
glinda
======

|ReadTheDocs| |Travis| |CodeClimate|

Glinda is a companion library for `tornado`_.  It is an attempt to make your
time with the framework less painful.  In fact, I want to make it downright
enjoyable.  I started down the path of developing HTTP endpoints in Tornado
and needing to test them.  The `tornado.testing`_ package is handy for
testing endpoints in isolation.  But what do you do when you have a HTTP
service that is calling other HTTP services asynchronously.  It turns out
that testing that is not as easy as it should be.  That is the first thing
that I tackled and it is the first thing that this library is going to
offer -- *a way to test non-trivial services*.

Once you can test your application, the next step is to write a well-behaved
application that fits into the WWW nicely.  Tornado does a pretty nice job
of handling the nitty gritty HTTP details (e.g., CTE, transfer encodings).
It doesn't provide a clean way to handle representations transparently so
I decided to add that into this library as well.

Content Handling
----------------
Tornado has some internal content decoding accessible by calling the
``get_body_arguments`` method of ``tornado.web.RequestHandler``.  It will
decode basic form data, ``application/x-www-form-urlencoded`` and
``multipart/form-data`` specifically.  Anything else is left up to you.
``glinda`` exposes a content handling mix-in that imbues a standard
``RequestHandler`` with a property that is the decoded request body and
a new method to encode a response.  Here's what it looks like:

.. code-block:: python

   class MyHandler(glinda.content.HandlerMixin, web.RequestHandler):
       def post(self, *args, **kwargs):
           body_argument = self.request_body['arg']
           # do stuff
           self.send_response(response_dict)
           self.finish()

   if __name__ == '__main__':
       glinda.content.register_text_type('application/json',
                                         default_charset='utf-8',
                                         dumper=json.dumps, loader=json.loads)
       glinda.content.register_binary_type('application/msgpack',
                                           msgpack.dumpb, msgpack.loadb)

When the client sends a post with a content type of ``application/json``, it
will decode the binary body to a string according to the HTTP headers and
call ``json.loads`` to decode the body when you reference the ``request_body``
property.  Failures are handled by raising a ``HTTPError(400)`` so you don't
have to worry about handling malformed messages.  The ``send_response``
method will take care of figuring out the appropriate content type based on
any included ``Accept`` headers.  All that you have to do is install
encoding and decoding handlers for expected content types.

The ``glinda.content`` package implements content handling as described in
`RFC7231`_.  Specifically, it decodes request bodies as described in
`section 3.1`_ and proactive content negotiation as described in sections
`3.4.1`_ and `5.3`_.

Testing
-------
Here's an example of testing a Tornado endpoint that asynchronously calls
another service.  In this case, the application interacts with  with the
``/add`` endpoint of some other service.  Testing in isolation can be tricky
without having to have a copy of the service running.  You could mock out
the ``AsyncHTTPClient`` and return fake futures and what not but that has
the nasty side-effect of hiding defects around how content type or headers
are handled -- no HTTP requests means that you have untested assumptions.

The following snippet tests the application under test using the
``ServiceLayer`` abstraction that ``glinda.testing`` provides.

.. code-block:: python

   from tornado import testing
   from glinda.testing import services


   class MyServiceTests(testing.AsyncHTTPTestCase):

      def setUp(self):
         service_layer = services.ServiceLayer()
         self.service = service_layer['adder']
         # TODO configured your application here using
         # self.service.url_for('/add') or self.service.host
         super(MyServiceTests, self).setUp()

      def get_app(self):
         return MyApplication()

      def test_that_my_service_calls_other_service(self):
         self.service.add_response(
            services.Request('POST', '/add'),
            services.Response(200, body='{"result": 10}'))
         self.fetch(self.get_url('/do-stuff'), method='GET')

         recorded = self.service.get_request('/add')
         self.assertEqual(recorded.method, 'POST')
         self.assertEqual(recorded.body, '[1,2,3,4]')
         self.assertEqual(recorded.headers['Content-Type'], 'application/json')

The application under test is linked in by implementing the standard
``tornado.testing.AsyncHTTPTestCase.get_app`` method.  Then you add in
a ``glinda.testing.services.ServiceLayer`` object and configure it to look
like the services that you depend on by adding endpoints and then configuring
your application to point at the service layer.  When you invoke the
application under test using ``self.fetch(...)``, it will send HTTP requests
through the Tornado stack (using the testing ``ioloop``) to the service layer
which will respond appropriately.  The beauty is that the entire HTTP stack is
exercised locally so that you can easily test edge cases such as correct
handling of status codes, custom headers, or malformed bodies without
resorting to deep mocking.

Where?
------
+---------------+-------------------------------------------------+
| Source        | https://github.com/dave-shawley/glinda          |
+---------------+-------------------------------------------------+
| Status        | https://travis-ci.org/dave-shawley/glinda       |
+---------------+-------------------------------------------------+
| Download      | https://pypi.python.org/pypi/glinda             |
+---------------+-------------------------------------------------+
| Documentation | http://glinda.readthedocs.org/en/latest         |
+---------------+-------------------------------------------------+
| Issues        | https://github.com/dave-shawley/glinda          |
+---------------+-------------------------------------------------+

.. _tornado: http://tornadoweb.org/
.. _tornado.testing: http://www.tornadoweb.org/en/latest/testing.html
.. _RFC7231: http://tools.ietf.org/html/rfc7231
.. _section 3.1: http://tools.ietf.org/html/rfc7231#section-3.1
.. _3.4.1: http://tools.ietf.org/html/rfc7231#section-3.4.1
.. _5.3: http://tools.ietf.org/html/rfc7231#section-5.3

.. |ReadTheDocs| image:: https://readthedocs.org/projects/glinda/badge/
   :target: https://glinda.readthedocs.org/
.. |Travis| image:: https://travis-ci.org/dave-shawley/glinda.svg
   :target: https://travis-ci.org/dave-shawley/glinda
.. |CodeClimate| image:: https://codeclimate.com/github/dave-shawley/glinda/badges/gpa.svg
   :target: https://codeclimate.com/github/dave-shawley/glinda


docs/changelog.rst
Changelog
---------

* `1.0.1`_ (27 Jun 2019)

  - Fix errant usage of :class:`tornado.web.ErrorHandler`

* `1.0.0`_ (22 Apr 2019)

  - Adjust supported Python versions

    - Add support for 3.5, 3.6, & 3.7
    - Drop support for 3.3, & 3.4

  - Adjust supported Tornado versions

    - Add support for 4.5, 5, and 6
    - Drop support for 3-4.5

* `0.1.0`_ (3 Jul 2017)

  - Modify content negotiation to 406 when asked for an unknown character set
  - Add :attr:`glinda.content.HandlerMixin.registered_content_types`
  - Add directory of examples
  - Remove support for tornado newer than 4.5.  *This is tempory due to changes
    in the Tornado API*.
  - Change Application.add_resource so that it inserts handlers before the
    default error handler instead of after.

* `0.0.3`_ (30 May 2015)

  - Add :func:`glinda.content.clear_handlers`
  - Add :func:`glinda.content.register_binary_type`
  - Add :func:`glinda.content.register_text_type`
  - Add :class:`glinda.content.HandlerMixin`

* `0.0.2`_ (29 May 2015)

  - Make testing layer actually return the headers and bodies that
    are programmatically added via ``glinda.testing.services.Response``.

* `0.0.1`_ (21 May 2015)

  - Add :class:`glinda.testing.services.Service`
  - Add :class:`glinda.testing.services.ServiceLayer`
  - Add :class:`glinda.testing.services.Request`
  - Add :class:`glinda.testing.services.Response`

.. _Next Release: https://github.com/dave-shawley/glinda/compare/1.0.1...master
.. _1.0.1: https://github.com/dave-shawley/glinda/compare/1.0.0...1.0.1
.. _1.0.0: https://github.com/dave-shawley/glinda/compare/0.1.0...1.0.0
.. _0.1.0: https://github.com/dave-shawley/glinda/compare/0.0.3...0.1.0
.. _0.0.3: https://github.com/dave-shawley/glinda/compare/0.0.2...0.0.3
.. _0.0.2: https://github.com/dave-shawley/glinda/compare/0.0.1...0.0.2
.. _0.0.1: https://github.com/dave-shawley/glinda/compare/0.0.0...0.0.1


tests/example_test.py
"""
Example of using glinda to test a service.

This file is an example of using the ``glinda.testing`` package to
test a very simple Tornado application.

"""
import os

from tornado import gen, httpclient, web
import tornado.testing

from glinda.testing import services


class MyHandler(web.RequestHandler):
    """
    Simple handler that makes a few asynchronous requests.

    The "backing" service is configured by setting the ``SERVICE_NETLOC``
    environment variable to the host and port of the service endpoint.
    This is pretty typical for 12 Factor Applications and it makes it
    very easy to test the service using ``glinda``.

    """

    @gen.coroutine
    def get(self):
        netloc = os.environ['SERVICE_NETLOC']
        client = httpclient.AsyncHTTPClient()
        response = yield client.fetch('http://{0}/status'.format(netloc),
                                      method='HEAD', raise_error=False)
        if response.code >= 300:
            raise web.HTTPError(504)

        response = yield client.fetch('http://{0}/do-stuff'.format(netloc),
                                      method='POST', raise_error=False,
                                      body='important stuff')
        if response.code >= 300:
            raise web.HTTPError(500)

        self.set_status(200)
        self.set_header('Custom', response.headers.get('Custom', ''))
        self.write(response.body)
        self.finish()


class HandlerTests(tornado.testing.AsyncHTTPTestCase):
    """Traditional :class:`tornado.testing.AsyncHTTPTestCase`"""

    def setUp(self):
        super(HandlerTests, self).setUp()
        service_layer = services.ServiceLayer()
        self.external_service = service_layer['status']
        os.environ['SERVICE_NETLOC'] = self.external_service.host

    def get_app(self):
        return web.Application([web.url('/do-the-things', MyHandler)])

    def test_that_status_failure_results_in_504(self):
        self.external_service.add_response(
            services.Request('HEAD', '/status'), services.Response(500))
        try:
            self.fetch('/do-the-things')
        except web.HTTPError as error:
            self.assertEqual(error.status_code, 504)

    def test_that_service_failure_results_in_500(self):
        self.external_service.add_response(
            services.Request('HEAD', '/status'), services.Response(200))
        self.external_service.add_response(
            services.Request('POST', '/do-stuff'), services.Response(400))
        try:
            self.fetch('/do-the-things')
        except web.HTTPError as error:
            self.assertEqual(error.status_code, 500)

    def test_that_status_is_fetched(self):
        self.external_service.add_response(
            services.Request('HEAD', '/status'), services.Response(200))
        self.external_service.add_response(
            services.Request('POST', '/do-stuff'), services.Response(200))
        self.fetch('/do-the-things')
        self.external_service.assert_request('HEAD', '/status')

    def test_that_stuff_is_posted(self):
        self.external_service.add_response(
            services.Request('HEAD', '/status'), services.Response(200))
        self.external_service.add_response(
            services.Request('POST', '/do-stuff'), services.Response(200))
        self.fetch('/do-the-things')

        request = self.external_service.get_request('do-stuff')
        self.assertEqual(request.body, b'important stuff')

    def test_that_post_response_is_preserved(self):
        self.external_service.add_response(
            services.Request('HEAD', '/status'), services.Response(200))
        self.external_service.add_response(
            services.Request('POST', '/do-stuff'),
            services.Response(200, body='foo', headers={'Custom': 'header'}))

        response = self.fetch('/do-the-things')
        self.assertEqual(response.body.decode(), 'foo')
        self.assertEqual(response.headers['Custom'], 'header')


tests/content_tests.py
import json
import re

from tornado import testing, web
import msgpack

from examples import contentneg
from glinda import content


KOREAN_TEXT = (u'\uc138\uacc4\ub97c \ud5a5\ud55c \ub300\ud654, '
               u'\uc720\ub2c8\ucf54\ub4dc\ub85c \ud558\uc2ed\uc2dc\uc624. '
               u'\uc81c10\ud68c \uc720\ub2c8\ucf54\ub4dc \uad6d\uc81c '
               u'\ud68c\uc758\uac00 1997\ub144 3\uc6d4 10\uc77c\ubd80\ud130 '
               u'12\uc77c\uae4c\uc9c0 \ub3c5\uc77c\uc758 \ub9c8\uc778\uc988'
               u'\uc5d0\uc11c \uc5f4\ub9bd\ub2c8\ub2e4. \uc9c0\uae08 \ub4f1'
               u'\ub85d\ud558\uc2ed\uc2dc\uc624. \uc774 \ud68c\uc758\uc5d0'
               u'\uc11c\ub294 \uc5c5\uacc4 \uc804\ubc18\uc758 \uc804\ubb38'
               u'\uac00\ub4e4\uc774 \ud568\uaed8 \ubaa8\uc5ec \ub2e4\uc74c'
               u'\uacfc \uac19\uc740 \ubd84\uc57c\ub97c \ub2e4\ub8f9\ub2c8'
               u'\ub2e4. - \uc778\ud130\ub137\uacfc \uc720\ub2c8\ucf54\ub4dc, '
               u'\uad6d\uc81c\ud654\uc640 \uc9c0\uc5ed\ud654, \uc6b4\uc601 '
               u'\uccb4\uc81c\uc640 \uc751\uc6a9 \ud504\ub85c\uadf8\ub7a8'
               u'\uc5d0\uc11c \uc720\ub2c8\ucf54\ub4dc\uc758 \uad6c\ud604, '
               u'\uae00\uaf34, \ubb38\uc790 \ubc30\uc5f4, \ub2e4\uad6d\uc5b4 '
               u'\ucef4\ud4e8\ud305.')


class JsonContentTests(testing.AsyncHTTPTestCase):

    def get_app(self):
        return web.Application([web.url('/', contentneg.HttpbinHandler)])

    def setUp(self):
        super(JsonContentTests, self).setUp()
        content.register_text_type('application/json', 'utf-8',
                                   json.dumps, json.loads)

    def tearDown(self):
        super(JsonContentTests, self).tearDown()
        content.clear_handlers()

    def test_that_response_uses_registered_encoder(self):
        response = self.fetch('/', headers={'Property': 'returned in body'})
        self.assertEqual(response.headers['Content-Type'],
                         'application/json; charset=utf-8')
        body = json.loads(response.body.decode('utf-8'))
        self.assertEqual(body['headers']['Property'], 'returned in body')

    def test_that_default_charset_is_honored(self):
        response = self.fetch('/', method='POST',
                              headers={'Content-Type': 'application/json'},
                              body=b'{"name":"Andr\xC3\xA9"}')
        self.assertEqual(response.headers['Content-Type'],
                         'application/json; charset=utf-8')
        body = json.loads(response.body.decode('utf-8'))
        self.assertEqual(body['body']['name'], u'Andr\u00E9')

    def test_that_body_decode_failure_results_in_client_error(self):
        response = self.fetch('/', method='POST',
                              headers={'Content-Type': 'application/json'},
                              body=b'not json')
        self.assertEqual(response.code, 400)

    def test_that_unknown_content_type_results_in_unsupp_media_type(self):
        response = self.fetch('/', method='POST',
                              headers={'Content-Type': 'application/xml'},
                              body=b'<?xml version="1.0"?><body/>')
        self.assertEqual(response.code, 415)


class ContentSelectionTests(testing.AsyncHTTPTestCase):

    def get_app(self):
        return web.Application([
            web.url('/', contentneg.HttpbinHandler),
            web.url('/rfc2295', contentneg.RFC2295Handler),
        ])

    def setUp(self):
        super(ContentSelectionTests, self).setUp()
        content.register_text_type('application/json', 'utf-8',
                                   json.dumps, json.loads)
        content.register_binary_type('application/msgpack', msgpack.packb,
                                     msgpack.unpackb)

    def tearDown(self):
        super(ContentSelectionTests, self).tearDown()
        content.clear_handlers()

    def test_that_simple_accept_header_is_honored(self):
        response = self.fetch('/', headers={'Accept': 'application/json'})
        self.assertEqual(response.headers['Content-Type'],
                         'application/json; charset=utf-8')
        body = json.loads(response.body.decode('utf-8'))
        self.assertEqual(body['headers']['Accept'], 'application/json')

        response = self.fetch('/', headers={'Accept': 'application/msgpack'})
        self.assertEqual(response.headers['Content-Type'],
                         'application/msgpack')
        body = msgpack.unpackb(response.body)
        self.assertEqual(body[b'headers'][b'Accept'], b'application/msgpack')

    def test_that_header_driven_translation_works(self):
        body = {'some': 'simple', 'and complex': ['body', 'elements']}
        response = self.fetch('/', method='POST', body=msgpack.packb(body),
                              headers={
                                  'Content-Type': 'application/msgpack',
                                  'Accept': 'application/json'})
        self.assertEqual(response.headers['Content-Type'],
                         'application/json; charset=utf-8')
        self.assertEqual(json.loads(response.body.decode('utf-8'))['body'],
                         body)

    def test_that_no_acceptable_content_type_raises_406(self):
        response = self.fetch('/', headers={'Accept': 'application/xml'})
        self.assertEqual(response.code, 406)

    def test_that_request_body_is_decoded_using_charset_parameter(self):
        body = u'{"value":"\u00A0\u00FF"}'
        headers = {
            'Content-Type': 'application/json; charset=latin1',
            'Accept': 'application/json',
            'Accept-Charset': 'utf8',
        }
        response = self.fetch('/', method='POST',
                              body=body.encode('latin1'),
                              headers=headers)
        self.assertEqual(response.code, 200)
        self.assertEqual(response.headers['Content-Type'],
                         'application/json; charset=utf8')
        self.assertEqual(json.loads(response.body.decode('utf-8'))['body'],
                         json.loads(body))

    def test_that_registered_types_are_available_to_handler(self):
        # see RFC2295 section 5
        alt_pattern = re.compile(
            r'^{"[^"]*"\s+\d+(\.\d+)?\s+(?P<properties>.*)}$')
        response = self.fetch('/rfc2295', headers={'Negotiate': 'vlist',
                                                   'Accept': 'plain/text'})
        self.assertEqual(response.code, 300)
        alternatives = []
        for type_info in response.headers['Alternatives'].split(','):
            match = alt_pattern.match(type_info.strip())
            for property in re.findall('{[^}]*}', match.group('properties')):
                name, value = property[1:-1].split(' ', 1)
                if name == 'type':
                    alternatives.append(value)
        self.assertEqual(sorted(alternatives),
                         ['application/json', 'application/msgpack'])


class TextEncodingTests(testing.AsyncHTTPTestCase):

    @staticmethod
    def unicode_dumper(body, **kwargs):
        kwargs['ensure_ascii'] = True
        return json.dumps(body, **kwargs)

    def get_app(self):
        return web.Application([web.url('/', contentneg.HttpbinHandler)])

    def setUp(self):
        super(TextEncodingTests, self).setUp()
        content.register_text_type('application/json', 'utf-8',
                                   self.unicode_dumper, json.loads)

    def tearDown(self):
        super(TextEncodingTests, self).tearDown()
        content.clear_handlers()

    def test_that_accept_charset_is_honored(self):
        str_body = json.dumps({'text': KOREAN_TEXT})
        encoded_body = str_body.encode('utf-8')
        response = self.fetch('/', body=encoded_body, method='POST', headers={
            'Content-Type': 'application/json; charset=utf-8',
            'Accept': 'application/json',
            'Accept-Charset': 'euc_kr',
        })
        self.assertEqual(response.headers['Content-Type'],
                         'application/json; charset=euc_kr')
        self.assertEqual(json.loads(response.body.decode('euc_kr'))['body'],
                         {'text': KOREAN_TEXT})

    def test_that_unknown_encoding_raises_406(self):
        response = self.fetch('/', headers={'Accept-Charset': 'foo'})
        self.assertEqual(response.code, 406)


tests/testing_tests.py
import unittest

from tornado import httpclient
import tornado.testing

from glinda import httpcompat
from glinda.testing import services


class ServiceUrlTests(unittest.TestCase):

    def setUp(self):
        super(ServiceUrlTests, self).setUp()
        service_layer = services.ServiceLayer()
        self.service = service_layer.get_service('service')

    def test_that_get_service_url_references_service(self):
        service_url = self.service.url_for()
        url = httpcompat.urlsplit(service_url)
        self.assertEqual(url.scheme, 'http')
        self.assertEqual(url.hostname, '127.0.0.1')
        self.assertNotEqual(url.port, 0)
        self.assertEqual(url.path, '/')
        self.assertEqual(url.query, '')
        self.assertEqual(url.fragment, '')

    def test_that_get_service_url_quotes_path(self):
        service_url = self.service.url_for('path that', 'needs', 'quo+ing')
        url = httpcompat.urlsplit(service_url)
        self.assertEqual(url.path.lower(), '/path%20that/needs/quo%2bing')

    def test_that_get_service_url_quotes_query(self):
        service_url = self.service.url_for(a='something with spaces',
                                           b='!@#$%^&*()')
        url = httpcompat.urlsplit(service_url)
        self.assertEqual(
            url.query.lower(),
            'a=something+with+spaces&b=%21%40%23%24%25%5e%26%2a%28%29'
        )

    def test_that_get_service_url_sorts_query(self):
        service_url = self.service.url_for(first=1, second=2, third=3,
                                           fini='first')
        url = httpcompat.urlsplit(service_url)
        self.assertEqual(url.query.lower(),
                         'fini=first&first=1&second=2&third=3')


class EndpointTests(tornado.testing.AsyncTestCase):

    def setUp(self):
        super(EndpointTests, self).setUp()
        self.service_layer = services.ServiceLayer()

    @tornado.testing.gen_test
    def test_that_endpoint_responds_with_456_by_default(self):
        service = self.service_layer['service']
        service.add_endpoint('resource')
        client = httpclient.AsyncHTTPClient()
        try:
            yield client.fetch(service.url_for('resource'))
        except httpclient.HTTPError as error:
            self.assertEqual(error.code, 456)

    @tornado.testing.gen_test
    def test_that_endpoint_responds_with_programmed_response(self):
        service = self.service_layer['service']
        service.add_response(services.Request('GET', '/resource:test'),
                             services.Response(222))

        client = httpclient.AsyncHTTPClient()
        response = yield client.fetch(service.url_for('resource:test'))
        self.assertEqual(response.code, 222)

    @tornado.testing.gen_test
    def test_that_empty_path_is_usable(self):
        service = self.service_layer['service']
        service.add_response(services.Request('GET', '/'),
                             services.Response(222))

        client = httpclient.AsyncHTTPClient()
        response = yield client.fetch(service.url_for('/'))
        self.assertEqual(response.code, 222)

    @tornado.testing.gen_test
    def test_that_error_handler_works(self):
        service = self.service_layer['service']

        client = httpclient.AsyncHTTPClient()
        response = yield client.fetch(service.url_for('/'), raise_error=False)
        self.assertEqual(response.code, 456)


class RequestRecordingTests(tornado.testing.AsyncTestCase):

    def setUp(self):
        super(RequestRecordingTests, self).setUp()
        self.service_layer = services.ServiceLayer()

    @tornado.testing.gen_test
    def test_that_all_request_details_are_recorded(self):
        service = self.service_layer['service']
        service.add_response(services.Request('POST', '/resource'),
                             services.Response(200))

        url = service.url_for('/resource', arg='value', arg2='value')
        client = httpclient.AsyncHTTPClient()
        yield client.fetch(url, method='POST', body='BODY',
                           headers={'Custom': 'Header'})
        req = next(service.get_requests_for('/resource'))
        self.assertEqual(req.method, 'POST')
        self.assertEqual(req.resource, '/resource')
        self.assertEqual(req.query, {'arg': 'value', 'arg2': 'value'})
        self.assertEqual(req.body, b'BODY')
        self.assertEqual(req.headers['Custom'], 'Header')
        self.assertIn('Host', req.headers)  # always present in HTTP/1.1

    def test_that_get_request_for_asserts_without_requests(self):
        service = self.service_layer['service']
        with self.assertRaises(AssertionError):
            next(service.get_requests_for('/whatever'))

    def test_that_get_request_asserts_without_requests(self):
        service = self.service_layer['service']
        with self.assertRaises(AssertionError):
            service.get_request('/whatever')

    @tornado.testing.gen_test
    def test_that_assert_request_fails_for_incorrect_request(self):
        service = self.service_layer['service']
        service.add_response(services.Request('GET', '/resource'),
                             services.Response(200))

        client = httpclient.AsyncHTTPClient()
        yield client.fetch(service.url_for('/resource', one='1', two=2))
        with self.assertRaises(AssertionError):
            service.assert_request('GET', '/resource')

    @tornado.testing.gen_test
    def test_that_assert_request_matches_all_parameters(self):
        service = self.service_layer['service']
        service.add_response(services.Request('GET', '/resource'),
                             services.Response(200))
        service.add_response(services.Request('POST', '/resource'),
                             services.Response(200))

        client = httpclient.AsyncHTTPClient()
        yield client.fetch(service.url_for('/resource', foo='bar'),
                           method='POST', body='')
        with self.assertRaises(AssertionError):
            service.assert_request('GET', '/resource', foo='bar')
        service.assert_request('POST', '/resource', foo='bar')


examples/contentneg.py
#!/usr/bin/env python
"""

Example tornado application that handles content negotiation.

This is an example of using glinda's content negotiation functionality
in a simple Tornado application.

"""
import importlib
import json
import logging
import os

from glinda import content, httpcompat
from ietfparse import headers
from tornado import httpserver, ioloop, web


# The following will be supported if available
def maybe_import(name):
    try:
        return importlib.import_module(name)
    except ImportError:
        return None

msgpack = maybe_import('msgpack')
yaml = maybe_import('yaml')


class HttpbinHandler(content.HandlerMixin, web.RequestHandler):
    """Mimics http://httpbin.org/{get,post}"""

    def get(self):
        self.send_response(self.standard_response_dict)
        self.finish()

    def post(self):
        response = self.standard_response_dict
        response['data'] = repr(self.request.body)
        response['files'] = {}
        response['form'] = {}
        response['body'] = self.get_request_body()
        self.send_response(response)
        self.finish()

    @property
    def standard_response_dict(self):
        return {
            'args': httpcompat.parse_qs(self.request.query),
            'headers': dict(self.request.headers),
            'origin': self.request.remote_ip,
            'url': self.request.uri,
        }


class RFC2295Handler(content.HandlerMixin, web.RequestHandler):
    """Implements a bare-bones version of RFC2295 Negotiation."""

    def prepare(self):
        super(RFC2295Handler, self).prepare()
        if not self._finished:
            negotiate = headers.parse_list_header(
                self.request.headers.get('Negotiate', ''))
            if 'vlist' in negotiate:
                variants = []
                for content_type in self.registered_content_types:
                    variants.append('{{"{0}" 1.0 {{type {1}}}}}'.format(
                        self.request.uri, content_type))
                self.set_header('Alternatives', ', '.join(variants))
                self.set_header('TCN', 'list')

    def send_response(self, response_dict):
        try:
            super(RFC2295Handler, self).send_response(response_dict)
        except web.HTTPError as error:
            if error.status_code == 406:
                self.set_status(300, 'Multiple Choices')
                self.set_header('Vary', 'negotiate, accept')
                return
            raise

    def get(self):
        self.send_response({'hi': 'there'})


if __name__ == '__main__':
    logging.basicConfig(level=logging.DEBUG,
                        format='%(levelname)1.1s %(name)s: %(message)s')
    content.register_text_type('application/json', 'utf-8',
                               dumper=json.dumps, loader=json.loads)
    if msgpack:
        content.register_binary_type('application/x-msgpack',
                                     dumper=msgpack.packb,
                                     loader=msgpack.unpackb)
    if yaml:
        content.register_text_type('application/yaml', 'utf-8',
                                   dumper=yaml.dump, loader=yaml.load)

    app = web.Application([
        web.url(r'/', HttpbinHandler),
        web.url(r'/negotiate', RFC2295Handler),
    ], debug=True)
    server = httpserver.HTTPServer(app)
    server.listen(int(os.environ.get('PORT', '8000')))
    iol = ioloop.IOLoop.instance()
    try:
        ioloop.IOLoop.instance().start()
        iol.start()
    except KeyboardInterrupt:
        logging.info('stopping application')
        iol.add_callback(iol.stop)


examples/testing.py
"""
Example of using glinda to test a service.

This file is an example of using the ``glinda.testing`` package to
test a very simple Tornado application.

"""
import os

from tornado import gen, httpclient, web
import tornado.testing

from glinda.testing import services


class MyHandler(web.RequestHandler):
    """
    Simple handler that makes a few asynchronous requests.

    The "backing" service is configured by setting the ``SERVICE_NETLOC``
    environment variable to the host and port of the service endpoint.
    This is pretty typical for 12 Factor Applications and it makes it
    very easy to test the service using ``glinda``.

    """

    @gen.coroutine
    def get(self):
        netloc = os.environ['SERVICE_NETLOC']
        client = httpclient.AsyncHTTPClient()
        try:
            yield client.fetch('http://{0}/status'.format(netloc),
                               method='HEAD')
        except web.HTTPError as error:
            if error.code >= 300:
                raise web.HTTPError(504)

        try:
            response = yield client.fetch('http://{0}/do-stuff'.format(netloc),
                                          method='POST',
                                          body='important stuff')
        except web.HTTPError as error:
            if error.code >= 300:
                raise web.HTTPError(500)

        self.set_status(200)
        self.set_header('Custom', response.headers.get('Custom', ''))
        self.write(response.body)
        self.finish()


class HandlerTests(tornado.testing.AsyncHTTPTestCase):
    """Traditional :class:`tornado.testing.AsyncHTTPTestCase`"""

    def setUp(self):
        super(HandlerTests, self).setUp()
        service_layer = services.ServiceLayer()
        self.external_service = service_layer['status']
        os.environ['SERVICE_NETLOC'] = self.external_service.host

    def get_app(self):
        return web.Application([web.url('/do-the-things', MyHandler)])

    def test_that_status_failure_results_in_504(self):
        self.external_service.add_response(
            services.Request('HEAD', '/status'), services.Response(500))
        try:
            self.fetch('/do-the-things')
        except web.HTTPError as error:
            self.assertEqual(error.status_code, 504)

    def test_that_service_failure_results_in_500(self):
        self.external_service.add_response(
            services.Request('HEAD', '/status'), services.Response(200))
        self.external_service.add_response(
            services.Request('POST', '/do-stuff'), services.Response(400))
        try:
            self.fetch('/do-the-things')
        except web.HTTPError as error:
            self.assertEqual(error.status_code, 500)

    def test_that_status_is_fetched(self):
        self.external_service.add_response(
            services.Request('HEAD', '/status'), services.Response(200))
        self.external_service.add_response(
            services.Request('POST', '/do-stuff'), services.Response(200))
        self.fetch('/do-the-things')
        self.external_service.assert_request('HEAD', '/status')

    def test_that_stuff_is_posted(self):
        self.external_service.add_response(
            services.Request('HEAD', '/status'), services.Response(200))
        self.external_service.add_response(
            services.Request('POST', '/do-stuff'), services.Response(200))
        self.fetch('/do-the-things')

        request = self.external_service.get_request('do-stuff')
        self.assertEqual(request.body, b'important stuff')

    def test_that_post_response_is_preserved(self):
        self.external_service.add_response(
            services.Request('HEAD', '/status'), services.Response(200))
        self.external_service.add_response(
            services.Request('POST', '/do-stuff'),
            services.Response(200, body='foo', headers={'Custom': 'header'}))

        response = self.fetch('/do-the-things')
        self.assertEqual(response.body.decode(), 'foo')
        self.assertEqual(response.headers['Custom'], 'header')
